File: photo-bridge/ios/PhotoBridge/Core/Protocol.swift
import Foundation

// Wire protocol mirror of docs/PROTOCOL.md (v1).
// Both this Swift implementation and the Rust receiver MUST pass the
// conformance vectors in conformance/.

public enum MsgType: UInt8, CaseIterable {
    case hello = 0
    case helloAck = 1
    case pinChallenge = 2
    case pinVerify = 3
    case pinOk = 4
    case pinFail = 5
    case sessionManifest = 6
    case fileState = 7
    case chunk = 8
    case ack = 9
    case nak = 10
    case fileDone = 11
    case fileVerified = 12
    case fileFailed = 13
    case sessionEnd = 14
    case sessionAbort = 15
    case ping = 16
    case pong = 17
}

public enum ProtocolError: Error, LocalizedError {
    case badMagic
    case oversize(Int)
    case truncated
    case emptyChunkHeader
    case invalidJSON
    case unexpectedType(MsgType)

    public var errorDescription: String? {
        switch self {
        case .badMagic: return "bad frame magic"
        case .oversize(let n): return "oversize frame: \(n)"
        case .truncated: return "truncated frame"
        case .emptyChunkHeader: return "empty chunk header"
        case .invalidJSON: return "invalid JSON payload"
        case .unexpectedType(let t): return "unexpected frame type \(t.rawValue)"
        }
    }
}

public enum ProtocolConstants {
    public static let maxPayload = 2 * 1024 * 1024
    public static let chunkSize = 1024 * 1024
    public static let window = 4
    public static let serviceType = "_photobridge._tcp"
    public static let port: UInt16 = 8471
}

public struct Frame {
    public let type: MsgType
    public let payload: Data
}

// MARK: - Payload types (JSON on the wire)

public struct Hello: Codable, Equatable {
    public var proto_ver: UInt32
    public var device_id: String
    public var device_name: String
    public var os: String
}

public struct HelloAck: Codable, Equatable {
    public var session_id: String
    public var paired: Bool
}

public struct PinChallenge: Codable, Equatable {
    public var challenge_id: String
    public var expires_at: Int64
}

public struct PinVerify: Codable, Equatable {
    public var challenge_id: String
    public var pin: String
}

public struct PinFail: Codable, Equatable {
    public var strikes_left: UInt32
}

public struct FileEntry: Codable, Equatable {
    public var file_id: String
    public var asset_id: String
    public var kind: String
    public var size: UInt64
    public var sha256: String
    public var created_at: Int64
    public var modified_at: Int64
    public var filename: String
}

public struct SessionManifest: Codable, Equatable {
    public var files: [FileEntry]
}

public enum FileStatus: String, Codable, Equatable {
    case new
    case partial
    case verified
    case dedup
}

public struct FileStateItem: Codable, Equatable {
    public var file_id: String
    public var status: FileStatus
    public var offset: UInt64
}

public struct FileState: Codable, Equatable {
    public var files: [FileStateItem]
}

public struct Ack: Codable, Equatable {
    public var file_id: String
    public var offset: UInt64
    public var bytes: UInt64
}

public struct Nak: Codable, Equatable {
    public var file_id: String
    public var offset: UInt64
    public var reason: String
}

public struct FileDone: Codable, Equatable {
    public var file_id: String
    public var sha256: String
}

public struct FileVerified: Codable, Equatable {
    public var file_id: String
    public var sha256: String
    public var size: UInt64
    public var dedup_of: String?
}

public struct FileFailed: Codable, Equatable {
    public var file_id: String
    public var reason: String
    public var attempt: UInt32
}

public struct SessionEnd: Codable, Equatable {
    public var files_total: UInt64
    public var bytes_total: UInt64
    public var files_done: UInt64
    public var bytes_done: UInt64
}

public struct SessionAbort: Codable, Equatable {
    public var reason: String
}

// Chunk header; first chunk of a file also carries "size" (informational,
// the receiver keys off its own manifest DB).
public struct ChunkHeader: Codable, Equatable {
    public var file_id: String
    public var offset: UInt64
    public var crc32: UInt32
    public var size: UInt64?
}

public enum ChunkData {
    case header(ChunkHeader)
    case payload(ChunkHeader, Data)
}

// MARK: - Framing codec

public enum Codec {
    public static let magic: [UInt8] = [0x50, 0x42]

    /// Frame = magic(2) + u32 BE length(4) + type(1) + payload.
    public static func encode(type: MsgType, payload: Data) throws -> Data {
        guard payload.count <= ProtocolConstants.maxPayload else {
            throw ProtocolError.oversize(payload.count)
        }
        var out = Data(magic)
        out.append(contentsOf: UInt32(payload.count).bytes)
        out.append(type.rawValue)
        out.append(payload)
        return out
    }

    public static func encode<T: Encodable>(type: MsgType, json: T) throws -> Data {
        try encode(type: type, payload: JSONEncoder.encodePB(json))
    }

    /// CHUNK payload: u32 BE header_len + JSON header + raw bytes.
    public static func chunkPayload(header: ChunkHeader, data: Data) throws -> Data {
        let hjson = try JSONEncoder.encodePB(header)
        guard 4 + hjson.count + data.count <= ProtocolConstants.maxPayload else {
            throw ProtocolError.oversize(4 + hjson.count + data.count)
        }
        var out = Data(UInt32(hjson.count).bytes)
        out.append(hjson)
        out.append(data)
        return out
    }

    /// Parse a CHUNK payload into (header, data).
    public static func splitChunk(_ payload: Data) throws -> ChunkData {
        guard payload.count >= 4 else { throw ProtocolError.emptyChunkHeader }
        let hlen = Int(payload.subdata(in: 0..<4).readBE32)
        guard hlen > 0, 4 + hlen <= payload.count else { throw ProtocolError.emptyChunkHeader }
        let hdr = try JSONDecoder.pb.decode(ChunkHeader.self, from: payload.subdata(in: 4..<(4 + hlen)))
        return .payload(hdr, payload.subdata(in: (4 + hlen)..<payload.count))
    }
}

// MARK: - Frame reader (incremental)

/// Buffered frame decoder: feed bytes, get frames.
public struct FrameReader {
    public private(set) var buffer = [UInt8]()
    private let maxFrame: Int = 7 + ProtocolConstants.maxPayload

    public init() {}

    public mutating func append(_ bytes: Data) {
        buffer.append(contentsOf: bytes)
        if buffer.count > maxFrame * 2 { buffer.removeFirst(buffer.count - maxFrame) }
    }

    /// Returns one complete frame if available.
    public mutating func nextFrame() throws -> Frame? {
        guard buffer.count >= 7 else { return nil }
        let magicOK = buffer[0] == Codec.magic[0] && buffer[1] == Codec.magic[1]
        guard magicOK else { throw ProtocolError.badMagic }
        let len = Int(UInt32(buffer[2]) << 24 | UInt32(buffer[3]) << 16 | UInt32(buffer[4]) << 8 | UInt32(buffer[5]))
        guard len <= ProtocolConstants.maxPayload else { throw ProtocolError.oversize(len) }
        guard buffer.count >= 7 + len else { return nil }
        let type = MsgType(rawValue: buffer[6])
        let payload = Data(buffer[7..<(7 + len)])
        buffer.removeFirst(7 + len)
        guard let type else { throw ProtocolError.unexpectedType(.init(rawValue: buffer[6]) ?? .ping) }
        return Frame(type: type, payload: payload)
    }
}

// MARK: - Small helpers

extension UInt32 {
    fileprivate var bytes: [UInt8] {
        [
            UInt8((self >> 24) & 0xFF), UInt8((self >> 16) & 0xFF),
            UInt8((self >> 8) & 0xFF), UInt8(self & 0xFF),
        ]
    }
}

extension Data {
    var readBE32: UInt32 {
        guard count >= 4 else { return 0 }
        return (UInt32(self[0]) << 24) | (UInt32(self[1]) << 16) | (UInt32(self[2]) << 8) | UInt32(self[3])
    }

    var hex: String { map { String(format: "%02x", $0) }.joined() }
}

extension JSONEncoder {
    static let pb: JSONEncoder = {
        let e = JSONEncoder()
        e.outputFormatting = []
        return e
    }()

    static func encodePB<T: Encodable>(_ value: T) throws -> Data {
        try pb.encode(value)
    }
}

extension JSONDecoder {
    static let pb = JSONDecoder()
}

File: photo-bridge/desktop/src-tauri/src/protocol.rs
use serde::{Deserialize, Serialize};
use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
use tokio::time::{timeout, Duration};

pub const MAGIC: [u8; 2] = [0x50, 0x42]; // "PB"
pub const MAX_PAYLOAD: usize = 2 * 1024 * 1024;
pub const CHUNK_SIZE: usize = 1024 * 1024; // 1 MiB
pub const WINDOW: usize = 4; // un-ACKed chunks in flight (sender side)
pub const HEARTBEAT_IDLE: Duration = Duration::from_secs(15);
pub const READ_TIMEOUT: Duration = Duration::from_secs(60);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[repr(u8)]
pub enum MsgType {
    Hello = 0,
    HelloAck = 1,
    PinChallenge = 2,
    PinVerify = 3,
    PinOk = 4,
    PinFail = 5,
    SessionManifest = 6,
    FileState = 7,
    Chunk = 8,
    Ack = 9,
    Nak = 10,
    FileDone = 11,
    FileVerified = 12,
    FileFailed = 13,
    SessionEnd = 14,
    SessionAbort = 15,
    Ping = 16,
    Pong = 17,
}

impl MsgType {
    pub fn from_u8(v: u8) -> Option<Self> {
        Some(match v {
            0 => Self::Hello,
            1 => Self::HelloAck,
            2 => Self::PinChallenge,
            3 => Self::PinVerify,
            4 => Self::PinOk,
            5 => Self::PinFail,
            6 => Self::SessionManifest,
            7 => Self::FileState,
            8 => Self::Chunk,
            9 => Self::Ack,
            10 => Self::Nak,
            11 => Self::FileDone,
            12 => Self::FileVerified,
            13 => Self::FileFailed,
            14 => Self::SessionEnd,
            15 => Self::SessionAbort,
            16 => Self::Ping,
            17 => Self::Pong,
            _ => return None,
        })
    }
}

#[derive(Debug, Clone)]
pub struct Frame {
    pub typ: MsgType,
    pub payload: Vec<u8>,
}

#[derive(Debug, thiserror::Error)]
pub enum CodecError {
    #[error("bad magic")]
    BadMagic,
    #[error("oversize frame: {0}")]
    Oversize(usize),
    #[error("truncated frame")]
    Truncated,
    #[error("unknown message type {0}")]
    UnknownType(u8),
    #[error("io error: {0}")]
    Io(#[from] std::io::Error),
    #[error("timed out")]
    Timeout,
    #[error("json error: {0}")]
    Json(#[from] serde_json::Error),
    #[error("db error: {0}")]
    Db(String),
    #[error("zero-length chunk header")]
    EmptyChunkHeader,
    #[error("protocol violation: {0}")]
    Violation(String),
}

pub fn encode_frame(typ: MsgType, payload: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity(7 + payload.len());
    out.extend_from_slice(&MAGIC);
    out.extend_from_slice(&(payload.len() as u32).to_be_bytes());
    out.push(typ as u8);
    out.extend_from_slice(payload);
    out
}

pub fn encode_json<T: Serialize>(typ: MsgType, val: &T) -> Result<Vec<u8>, CodecError> {
    let payload = serde_json::to_vec(val)?;
    Ok(encode_frame(typ, &payload))
}

/// Payload for a CHUNK frame: u32 BE header_len + JSON header + raw bytes.
pub fn encode_chunk(file_id: &str, offset: u64, crc32: u32, data: &[u8]) -> Result<Vec<u8>, CodecError> {
    let hdr = ChunkHeader { file_id: file_id.to_string(), offset, crc32 };
    let hdr_json = serde_json::to_vec(&hdr)?;
    let payload_len = 4 + hdr_json.len() + data.len();
    let mut payload = Vec::with_capacity(payload_len);
    payload.extend_from_slice(&(hdr_json.len() as u32).to_be_bytes());
    payload.extend_from_slice(&hdr_json);
    payload.extend_from_slice(data);
    Ok(encode_frame(MsgType::Chunk, &payload))
}

/// Read one complete frame from the stream (with timeout).
pub async fn read_frame<R: AsyncRead + Unpin>(r: &mut R) -> Result<Frame, CodecError> {
    let mut header = [0u8; 7];
    timeout(READ_TIMEOUT, r.read_exact(&mut header))
        .await
        .map_err(|_| CodecError::Timeout)?
        .map_err(|e| {
            if e.kind() == std::io::ErrorKind::UnexpectedEof {
                CodecError::Truncated
            } else {
                CodecError::Io(e)
            }
        })?;
    if header[0] != MAGIC[0] || header[1] != MAGIC[1] {
        return Err(CodecError::BadMagic);
    }
    let len = u32::from_be_bytes([header[2], header[3], header[4], header[5]]) as usize;
    if len > MAX_PAYLOAD {
        return Err(CodecError::Oversize(len));
    }
    let typ = MsgType::from_u8(header[6]).ok_or(CodecError::UnknownType(header[6]))?;
    let mut payload = vec![0u8; len];
    timeout(READ_TIMEOUT, r.read_exact(&mut payload))
        .await
        .map_err(|_| CodecError::Timeout)?
        .map_err(|e| {
            if e.kind() == std::io::ErrorKind::UnexpectedEof {
                CodecError::Truncated
            } else {
                CodecError::Io(e)
            }
        })?;
    Ok(Frame { typ, payload })
}

pub async fn write_frame<W: AsyncWrite + Unpin>(w: &mut W, typ: MsgType, payload: &[u8]) -> Result<(), CodecError> {
    let frame = encode_frame(typ, payload);
    w.write_all(&frame).await?;
    Ok(())
}

pub async fn write_json<W: AsyncWrite + Unpin, T: Serialize>(w: &mut W, typ: MsgType, val: &T) -> Result<(), CodecError> {
    write_frame(w, typ, &serde_json::to_vec(val)?).await
}

/// Split a CHUNK payload into (header, data).
pub fn split_chunk_payload(payload: &[u8]) -> Result<(ChunkHeader, &[u8]), CodecError> {
    if payload.len() < 4 {
        return Err(CodecError::EmptyChunkHeader);
    }
    let hlen = u32::from_be_bytes([payload[0], payload[1], payload[2], payload[3]]) as usize;
    if hlen == 0 || 4 + hlen > payload.len() {
        return Err(CodecError::EmptyChunkHeader);
    }
    let hdr: ChunkHeader = serde_json::from_slice(&payload[4..4 + hlen])?;
    Ok((hdr, &payload[4 + hlen..]))
}

// ---------------------------------------------------------------------------
// Payload types (JSON)
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChunkHeader {
    pub file_id: String,
    pub offset: u64,
    pub crc32: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Hello {
    pub proto_ver: u32,
    pub device_id: String,
    pub device_name: String,
    pub os: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HelloAck {
    pub session_id: String,
    pub paired: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PinChallenge {
    pub challenge_id: String,
    pub expires_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PinVerify {
    pub challenge_id: String,
    pub pin: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PinFail {
    pub strikes_left: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileEntry {
    pub file_id: String,
    pub asset_id: String,
    pub kind: String,
    pub size: u64,
    pub sha256: String,
    pub created_at: i64,
    pub modified_at: i64,
    pub filename: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionManifest {
    pub files: Vec<FileEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum FileStatus {
    New,
    Partial,
    Verified,
    Dedup,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileStateItem {
    pub file_id: String,
    pub status: FileStatus,
    pub offset: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileState {
    pub files: Vec<FileStateItem>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Ack {
    pub file_id: String,
    pub offset: u64,
    pub bytes: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Nak {
    pub file_id: String,
    pub offset: u64,
    pub reason: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileDone {
    pub file_id: String,
    pub sha256: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileVerified {
    pub file_id: String,
    pub sha256: String,
    pub size: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dedup_of: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileFailed {
    pub file_id: String,
    pub reason: String,
    pub attempt: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionEnd {
    pub files_total: u64,
    pub bytes_total: u64,
    pub files_done: u64,
    pub bytes_done: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionAbort {
    pub reason: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn frame_roundtrip() {
        let msg = Hello { proto_ver: 1, device_id: "d1".into(), device_name: "iPhone".into(), os: "ios".into() };
        let bytes = encode_json(MsgType::Hello, &msg).unwrap();
        let mut cursor = std::io::Cursor::new(bytes);
        let rt = tokio::runtime::Runtime::new().unwrap();
        let frame = rt.block_on(async { read_frame(&mut cursor).await }).unwrap();
        assert_eq!(frame.typ, MsgType::Hello);
        let parsed: Hello = serde_json::from_slice(&frame.payload).unwrap();
        assert_eq!(parsed.device_id, "d1");
    }

    #[test]
    fn chunk_roundtrip() {
        let data = vec![0xABu8; 1024];
        let bytes = encode_chunk("file-1", 0, 123456, &data).unwrap();
        let mut cursor = std::io::Cursor::new(bytes);
        let rt = tokio::runtime::Runtime::new().unwrap();
        let frame = rt.block_on(async { read_frame(&mut cursor).await }).unwrap();
        assert_eq!(frame.typ, MsgType::Chunk);
        let (hdr, body) = split_chunk_payload(&frame.payload).unwrap();
        assert_eq!(hdr.file_id, "file-1");
        assert_eq!(hdr.offset, 0);
        assert_eq!(hdr.crc32, 123456);
        assert_eq!(body, data.as_slice());
    }

    #[test]
    fn bad_magic_rejected() {
        let mut bytes = encode_frame(MsgType::Ping, b"{}").to_vec();
        bytes[0] = 0x00;
        let mut cursor = std::io::Cursor::new(bytes);
        let rt = tokio::runtime::Runtime::new().unwrap();
        let err = rt.block_on(async { read_frame(&mut cursor).await }).unwrap_err();
        assert!(matches!(err, CodecError::BadMagic));
    }

    #[test]
    fn oversize_rejected() {
        let mut bytes = encode_frame(MsgType::Ping, b"{}").to_vec();
        bytes[2..6].copy_from_slice(&(MAX_PAYLOAD as u32 + 1).to_be_bytes());
        let mut cursor = std::io::Cursor::new(bytes);
        let rt = tokio::runtime::Runtime::new().unwrap();
        let err = rt.block_on(async { read_frame(&mut cursor).await }).unwrap_err();
        assert!(matches!(err, CodecError::Oversize(_)));
    }
}